pkg/auth/keyring.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package auth
+
+import (
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	defaultSessionName = "_ses.nydus"
+)
+
+type KeyRing struct {
+	lock sync.RWMutex
+}
+
+func NewKeyRing() *KeyRing {
+	return &KeyRing{}
+}
+
+func (k *KeyRing) Add(id, value string) (int, error) {
+	sessKeyID, err := unix.KeyctlJoinSessionKeyring(defaultSessionName)
+	if err != nil {
+		return 0, errors.Wrap(err, "unable to session key")
+	}
+
+	if err := k.modifyKeyringPerm(sessKeyID, 0xffffffff, 0x80000); err != nil {
+		return 0, errors.Wrap(err, "unable to mod keyring permissions")
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	keyID, err := unix.AddKey("user", id, []byte(value), sessKeyID)
+	if err != nil {
+		return 0, err
+	}
+
+	return keyID, nil
+}
+
+func (k *KeyRing) modifyKeyringPerm(ringID int, mask, setbits uint32) error {
+	dest, err := unix.KeyctlString(unix.KEYCTL_DESCRIBE, ringID)
+	if err != nil {
+		return err
+	}
+
+	res := strings.Split(dest, ";")
```

Comment:
It's better to add a output example as comment to helper code readers to understand

Reply:
Done.